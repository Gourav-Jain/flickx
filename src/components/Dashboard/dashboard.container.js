import { useCallback, useEffect, useState } from "react";
import { debounce } from 'lodash';
import { useDispatch } from "react-redux";
import {
    useSelectorHook, getSearchMovieApiUrl, getPopularMovieApiUrl,
    updatedFavPayload, updatedWlPayload,
    API_KEY, API_SERACH_MOVIE_URL, API_POPULAR_MOVIE_URL
} from '../../utils'
import {
    setSearchMoviesRequest, setSearchMoviesResponse, setAddToFavourites,
    setAddToWatchLater, setRemoveFromFavourite, setRemoveFromWatchLater, setSearchMoviesFailer
} from "../../features/actions";

export const useDashboardContainer = () => {
    const { searchMovies } = useSelectorHook();
    const [errorMessage, setErrorMessage] = useState('')
    const dispatch = useDispatch()

    useEffect(() => {
        fetchMovies('');
    }, [])

    const fetchMovies = (value) => {
        const payload = { showLoader: true }
        dispatch(setSearchMoviesRequest(payload));
        const url = value ? getSearchMovieApiUrl(API_SERACH_MOVIE_URL, API_KEY, value) : getPopularMovieApiUrl(API_POPULAR_MOVIE_URL, API_KEY);
        //const url = `http://localhost:3001/search/movies/?q=${value}`;
        fetch(url, { method: "GET" }).then(data => {

            if (data.status === 200) {

                data.json().then(
                    res => {
                        if (res.results.length > 0) {
                            const payload = {
                                showLoader: false,
                                isSearch: value ? true : false,
                                searchResult: { result: res["results"] },
                            };

                            setErrorMessage('');
                            dispatch(setSearchMoviesResponse(payload));
                        } else {
                            const payload = {
                                error: true,
                                showLoader: false,
                            };
                            setErrorMessage('No movies found...');
                            dispatch(setSearchMoviesFailer(payload));
                        }
                    }
                )
            } else {
                const payload = {
                    error: true,
                    showLoader: false,
                };
                setErrorMessage('Opps, something went wrong... :(');
                dispatch(setSearchMoviesFailer(payload));
            }
        }).catch((error) => {
            if (error) {
                const payload = {
                    error: true,
                    showLoader: false,
                };
                setErrorMessage(error);
                dispatch(setSearchMoviesFailer(payload));
            }
        });
    }

    const debQry = useCallback(debounce((value) => fetchMovies(value), 500), []);

    const handleChange = e => {
        debQry(e.target.value);
    };


    // handle add to favourits list
    const handleAddToFav = (id) => {
        const payload = updatedFavPayload("fav", "add", id, searchMovies);
        dispatch(setAddToFavourites(payload));
    }

    // handle remove from favourits list
    const handleRemoveAddToFave = (id) => {
        const payload = updatedFavPayload("fav", "rmv", id, searchMovies);
        dispatch(setRemoveFromFavourite(payload));
    }

    // handle add to watch later list
    const handleAddToWatchLater = (id) => {
        const payload = updatedWlPayload("wl", "add", id, searchMovies);
        dispatch(setAddToWatchLater(payload));
    }

    // handle remove watch later list
    const handleRemoveWatchLater = (id) => {

        const payload = updatedWlPayload("wl", "rmv", id, searchMovies);
        dispatch(setRemoveFromWatchLater(payload));
    }

    return {
        showLoader: searchMovies.showLoader || false,
        error: searchMovies.error || false,
        errorMessage,
        handleChange,
        fetchMovies,
        handleAddToFav,
        handleAddToWatchLater,
        handleRemoveAddToFave,
        handleRemoveWatchLater,
        ...searchMovies.searchResult,
        favourites: searchMovies.favourites || [],
        watchLater: searchMovies.watchLater || [],
    };
};