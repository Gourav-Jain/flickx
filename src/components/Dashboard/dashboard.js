import React from 'react';
import { Input, Button, Segment, Popup, Message } from 'semantic-ui-react';
import { useDashboardContainer } from './dashboard.container'
import { Movies } from '../Movies/movies'
import { Footer } from '../Footer/footer'

export const Dashboard = () => {

    const { showLoader,
        error,
        errorMessage,
        handleChange,
        handleAddToFav,
        handleRemoveAddToFave,
        handleAddToWatchLater,
        handleRemoveWatchLater,
        result,
        favourites,
        watchLater } = useDashboardContainer();

    return (
        <>
            <Segment clearing>

                <div>
                    <Popup content="Favorites" trigger={
                        <Button
                            floated='right'
                            color='yellow'
                            content=''
                            icon='favorite'
                            label={{ basic: true, color: 'black', pointing: 'left', content: favourites.length }}
                        />}
                    />
                    <Popup content="Watch Later" trigger={<Button
                        color='blue'
                        content=''
                        icon='time'
                        floated='right'
                        label={{
                            as: 'a',
                            basic: true,
                            color: 'black',
                            pointing: 'left',
                            content: watchLater.length,
                        }}
                    />}
                    />
                </div>
                <Input floated='right' loading={showLoader} size='large' icon='search' placeholder='Search...' onChange={handleChange} />
            </Segment>
            {
                error && errorMessage ? <Message error={error}>
                    <p>{errorMessage}</p>
                </Message> : <Movies result={result} addToFavo={handleAddToFav} addToWatchLater={handleAddToWatchLater} showLoader={showLoader} />
            }


            <Footer favourites={favourites} removeFavourite={handleRemoveAddToFave} watchLater={watchLater} removeWatchLater={handleRemoveWatchLater} />
        </>
    )
}

export default Dashboard;