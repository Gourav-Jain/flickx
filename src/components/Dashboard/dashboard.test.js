import React from "react";
import { shallow } from "enzyme";

import Dashboard from "./dashboard";
import * as data from './dashboard.container';

describe("<Dashboard />", () => {

    const mockData = {
        showLoader: false,
        handleChange: jest.fn(),
        handleAddToFav: jest.fn(),
        handleRemoveAddToFave: jest.fn(),
        handleAddToWatchLater: jest.fn(),
        handleRemoveWatchLater: jest.fn(),
        result: [],
        favourites: [],
        watchLater: []
    };

    const useDashboardContainer = jest
        .spyOn(data, "useDashboardContainer")
        .mockReturnValue(mockData);

    beforeEach(() => {
        jest.clearAllMocks();

        useDashboardContainer.mockReturnValue(mockData);
    });

    describe("@actions", () => {
        test("should display Dashboard", () => {
            const dashboard = shallow(<Dashboard />);
            expect(dashboard).toMatchSnapshot();
        });
    });
});