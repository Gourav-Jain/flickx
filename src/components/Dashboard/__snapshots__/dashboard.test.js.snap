// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> @actions should display Dashboard 1`] = `ShallowWrapper {}`;
